### DISCUSSION TITLE
Pycharm warning on add_middleware with CORSMiddleware

### DISCUSSION THREAD

**Discussion by halon176** [fastapi/fastapi#10968] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
from fastapi.middleware.cors import CORSMiddleware

app = FastAPI()

origins = [
    "http://localhost.tiangolo.com",
    "https://localhost.tiangolo.com",
    "http://localhost",
    "http://localhost:8080",
]

app.add_middleware(
    CORSMiddleware,
    allow_origins=origins,
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)


@app.get("/")
async def main():
    return {"message": "Hello World"}
```

### Description

After updating to version 0.109, the indicated syntax generates a type warning in PyCharm (which does not occur in VSCode). The warning is as follows:

`Expected type 'Type[_MiddlewareClass]' (matched generic type 'Type[_MiddlewareClass[ParamSpec("P")]]'), got 'Type[CORSMiddleware]' instead.`

I have observed that there have been updates on the Starlet side, but I cannot confirm if this is impactful. For now, everything seems to be functioning correctly; the only issue is the IDE warning.

### Operating System

Linux

### Operating System Details

ubuntu 22.04

### FastAPI Version

0.109.0

### Pydantic Version

2.5.3

### Python Version

3.12.1

### Additional Context

PyCharm 2023.2.5 (Community Edition)

**Comment by aanchlia** [comment]
I think it's related to starlette version `0.35.0`.
https://github.com/encode/starlette/pull/2381

**Reply by rkrell** [reply]
The implementation works, but this typing warning by PyCharm is also correct.
In fact there is an internal mismatch in typing here, from my point of view, there isn't any _MiddlewareClass used to inherit from in the built-in midleware classes.

So, this one is the more particular and better answer.

**Reply by trw226** [reply]
The same issue occurs with SessionMiddleware.

**Reply by Kludex** [reply]
> In fact there is an internal mismatch in typing here, from my point of view, there isn't any _MiddlewareClass used to inherit from in the built-in midleware classes.

This is not accurate. The type hints are correct on Starlette's side.

**Comment by Kludex** [comment, marked as answer]
I recommend checking the PyCharm bug track.

The implementation looks fine.

**Reply by matpag** [reply]
Hi @Kludex, sorry to bother you but it doesn't seem a bug of PyCharm, can you read the comment above here: https://github.com/fastapi/fastapi/discussions/10968#discussioncomment-8230301?

Thanks

**Reply by Kludex** [reply]
If it doesn't reproduce with `pyright` and `mypy`, it's a PyCharm issue.

**Reply by Kludex** [reply]
PyCharm doesn't understand `ParamSpec` correctly.

**Reply by matpag** [reply]
Thanks for the additional info 🙏

**Reply by Kludex** [reply]
No worries.

**Comment by zystudios** [comment]
fastapi==0.112.2  also has this problem in pycharm
![image](https://github.com/user-attachments/assets/ffcf95a6-e570-414f-8a00-24e64c5bc32e)

**Comment by thelokeshgoel00** [comment]
I have raised a bug for PyCharm Team. https://youtrack.jetbrains.com/issue/PY-76760/FASTAPI-CorsMiddleware-type-issue .

I have referenced your reply @Kludex

**Reply by Kludex** [reply]
Thanks. Historically, PyCharm has been a pain as a maintainer.

My life would be easier if they open a channel to talk to the maintainers.

**Reply by etoricky** [reply]
Hit the same Pycharm warning. Expect Pycharm would have fixed it soon.

**Reply by 0xb1b1** [reply]
> Hit the same Pycharm warning. Expect Pycharm would have fixed it soon.

We're still waiting, by the way :)

**Comment by UnoYakshi** [comment]
[ty](https://github.com/astral-sh/ty/) (type checker from Astral) also marks it as an error (`invalid-argument-type`).

For this code:
```python
# Imports and stuff...

app = FastAPI(**app_configs, lifespan=lifespan)

# Add middlewares, mounts, routers, etc...
app.add_middleware(
    middleware_class=CORSMiddleware,
    allow_origins=settings.CORS_ORIGINS,
    allow_credentials=settings.CORS_CREDENTIALS,
    allow_methods=settings.CORS_METHODS,
    allow_headers=settings.CORS_HEADERS,
)
```

after I run ty:
```shell
uv run ty check .
```

I get:
```
error[invalid-argument-type]: Argument to bound method `add_middleware` is incorrect
  --> src/core/main.py:41:5
   |
39 | # Add middlewares, mounts, routers, etc...
40 | app.add_middleware(
41 |     middleware_class=CORSMiddleware,
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Expected `_MiddlewareFactory[Unknown]`, found `<class 'CORSMiddleware'>`
42 |     allow_origins=settings.CORS_ORIGINS,
43 |     allow_credentials=settings.CORS_CREDENTIALS,
   |
info: Method defined here
   --> .venv/lib/python3.13/site-packages/starlette/applications.py:124:9
    |
122 |         self.router.host(host, app=app, name=name)  # pragma: no cover
123 |
124 |     def add_middleware(
    |         ^^^^^^^^^^^^^^
125 |         self,
126 |         middleware_class: _MiddlewareFactory[P],
    |         --------------------------------------- Parameter declared here
127 |         *args: P.args,
128 |         **kwargs: P.kwargs,
    |
info: rule `invalid-argument-type` is enabled by default
```

I've checked the implementation of `fastapi.middleware.cors.CORSMiddleware`:
```python
class CORSMiddleware:
    def __init__(
        self,
        app: ASGIApp,
        # ...
```

It doesn't seem to explicitly state it is anyhow related to `_MiddlewareFactory[P]` of `starlette.applications.Starlette.add_middleware()`'s signature:
```python
    def add_middleware(
        self,
        middleware_class: _MiddlewareFactory[P],
        *args: P.args,
        **kwargs: P.kwargs,
    ) -> None:
    ...
```

**Reply by jvdavim** [reply]
Not sure if it's a FastAPI, Starlette or ty (from astral) issue, but ty (0.0.9) is yet showing as type error.

**Reply by UnoYakshi** [reply]
@jvdavim, I presume it is at least on FastAPI's side, because:
1. I took the implementation of `fastapi.middlware.cors.CORSMiddlware`.
2. `add_middleware()` expects `_MiddlewareFactory[P]`. I don't think it's connect to the current implementation anyhow.

**Reply by Kludex** [reply]
Ask on `ty`.

**Reply by norbusan** [reply]
Asked at ty: https://github.com/astral-sh/ty/issues/2406

**Reply by UnoYakshi** [reply]
@norbusan, thanks! Seems to be a [known issue](https://github.com/astral-sh/ty/issues/1635), however.

**Reply by norbusan** [reply]
Yup, so it will hopefully soon be fixed
